src/rosdiscover/recover/symbolic.py

```diff
@@ -146,6 +153,147 @@ def to_dict(self) -> t.Dict[str, t.Any]:
+class ThisExpr(SymbolicExpr, abc.ABC):
+
+    """Represents a symbolic value in a function summary."""
+    def to_dict(self) -> t.Dict[str, t.Any]:
+        return {
+            "kind": "ThisExpr",
+        }
+
+    def eval(self, context: SymbolicContext) -> t.Any:
+        return "this"
+
+    def to_str(self) -> str:
+        return "this"
+
+
+@attr.s(auto_attribs=True, slots=True)
+class NullExpr(SymbolicExpr, abc.ABC):
+
+    """Represents a symbolic value in a function summary."""
+    def to_dict(self) -> t.Dict[str, t.Any]:
+        return {
+            "kind": "NullExpr",
+        }
+
+    def eval(self, context: SymbolicContext) -> t.Any:
+        return "NULL"
+
+    def to_str(self) -> str:
+        return "NULL"
+
+
+@attr.s(auto_attribs=True, slots=True)
+class NegateExpr(SymbolicExpr, abc.ABC):
+    sub_expr: SymbolicExpr
+
+    """Represents a symbolic value in a function summary."""
+    def to_dict(self) -> t.Dict[str, t.Any]:
+        return {
+            "kind": "NegateExpr",
+            "subExpr": self.sub_expr.to_dict(),
+        }
+
+    def eval(self, context: SymbolicContext) -> t.Any:
+        return "not self.sub_expr.eval(context)"
+
+    def to_str(self) -> str:
+        return f"!{self.sub_expr.to_str()}"
+
+
+@attr.s(auto_attribs=True, slots=True)
+class BinaryExpr(SymbolicExpr, abc.ABC):
+    lhs: SymbolicExpr
+    rhs: SymbolicExpr
+
+    @abc.abstractmethod
+    def binary_operator(self) -> str:
+        ...
+
+    """Represents a symbolic value in a function summary."""
+    def to_dict(self) -> t.Dict[str, t.Any]:
+        return {
+            "kind": "BinaryExpr",
+            "operator": self.binary_operator(),
+            "lhs": self.lhs.to_dict(),
+            "rhs": self.rhs.to_dict(),
+        }
+
+    def to_str(self) -> str:
+        return f"{self.lhs.to_str()} {self.binary_operator()} {self.rhs.to_str()}"
+
+
+@attr.s(auto_attribs=True, slots=True)
+class CompareExpr(BinaryExpr, abc.ABC):
+    operator: str
+
+    def binary_operator(self) -> str:
+        return self.operator
+
+    def eval(self, context: SymbolicContext) -> t.Any:
+        if operator == "<=":
+            return self.lhs.eval(context) <= self.rhs.eval(context)
+        elif operator == "<":
+            return self.lhs.eval(context) < self.rhs.eval(context)
+        elif operator == ">=":
+            return self.lhs.eval(context) >= self.rhs.eval(context)
+        elif operator == ">":
+            return self.lhs.eval(context) > self.rhs.eval(context)
+        elif operator == "==":
+            return self.lhs.eval(context) == self.rhs.eval(context)
+        elif operator == "!=":
+            return self.lhs.eval(context) != self.rhs.eval(context)
+
+
+@attr.s(auto_attribs=True, slots=True)
+class BinaryMathExpr(BinaryExpr, abc.ABC):
+    operator: str
+
+    def binary_operator(self) -> str:
+        return self.operator
+
+    def eval(self, context: SymbolicContext) -> t.Any:
+        if operator == "+":
```

Comment:
:bug: see above

Reply:
fixed

Reply:
this still doesn't appear to be fixed? (GitHub could be misbehaving)

Reply:
now actually fixed